### PR TITLE
Fixes for latest GeoInterface.jl

### PR DESCRIPTION
**Pull request by ErickChacon**
GeoInterface integration using curren suggested approach and avoid using `GI.coordinates` to reduce problemas like in issue JuliaEarth/GeoTables.jl#14. This current PR fails with `Shapefile.jl` when working with `LineStrings` and `Polygons` because the method `GI.is3d` is not working (but a PR has already been done) for this geometries. The reason is that Shapefile reads `LineString` as `MultiLineString` and `Polygon` as `MultiPolygon`.

### CODE REVIEW

**Comment by ErickChacon** [comment]
@juliohm I started the pull request to agree in this GeoInterface integration. As mentioned at the beginning of this PR, it is not working with `Shapefile.jl` basically because the `GI.is3d` method is not working for `LineString` and `Polygon` because they are subparts of `MultiLineString` and `MultiPolygon` (for this geometries GI.is3d works fine).

Due to this issue with `Shapefile.jl` and also that we will call `GI.is3d` several times for multipoints, mutlines, multigeomtries. do you think is better to write a function that calls `GI.is3d` only once? I am already doing that using some auxiliary functions.

File: src/conversion.jl [hunk 1]
```diff
@@ -40,33 +41,70 @@ GI.getfeature(::Any, d::Data, i) = d[i,:]
...
+    Multi([GI.convert(Point, GI.PointTrait(), g) for g in GI.getgeom(geom)])
+end
+
+function GI.convert(::Type{Multi}, ::GI.MultiLineStringTrait, geom)
+    Multi([GI.convert(Chain, GI.LineStringTrait(), g) for g in GI.getgeom(geom)])
+end
+
+function GI.convert(::Type{Multi}, ::GI.MultiPolygonTrait, geom)
+    Multi([GI.convert(Polygon, GI.PolygonTrait(), g) for g in GI.getgeom(geom)])
+end
```
Comment (juliohm) [inline src/conversion.jl:95]:
```suggestion
function GI.convert(::Type{Multi}, ::GI.MultiPointTrait, geom)
  Multi([GI.convert(Point, GI.PointTrait(), g) for g in GI.getgeom(geom)])
end

function GI.convert(::Type{Multi}, ::GI.MultiLineStringTrait, geom)
  Multi([GI.convert(Chain, GI.LineStringTrait(), g) for g in GI.getgeom(geom)])
end

function GI.convert(::Type{Multi}, ::GI.MultiPolygonTrait, geom)
  Multi([GI.convert(Polygon, GI.PolygonTrait(), g) for g in GI.getgeom(geom)])
end
```

File: src/conversion.jl [hunk 2]
```diff
@@ -40,33 +41,70 @@ GI.getfeature(::Any, d::Data, i) = d[i,:]
...
 
-coords2multichain(coords) = Multi(coords2chain.(coords))
+function GI.convert(::Type{Segment}, ::GI.LineTrait, geom)
+  p1, p2 = GI.startpoint(geom), GI.endpoint(geom)
+  if GI.is3d(geom)
+    Segment(Point(GI.x(p1), GI.y(p1), GI.z(p1)), Point(GI.x(p2), GI.y(p2), GI.z(p1)))
+  else
+    Segment(Point(GI.x(p1), GI.y(p1)), Point(GI.x(p2), GI.y(p2)))
+  end
```
Comment (juliohm) [inline src/conversion.jl:66]: Maybe we could use `Segment(getpoints(geom))` here as well?

**Review by juliohm — CHANGES_REQUESTED** [review]
That is an awesome PR @ErickChacon, thanks for working on it. I understand that we need to wait for the change in Shapefile.jl to get merged and released before merging and releasing this one right?

**Comment by ErickChacon** [comment]
@juliohm, sorry. I just made another push. I changed the approach to call `GI.is3d` only once for each conversion. Please review the updated version and let me know if you prefer the initial submission. The advantage of this modification is that it already works with `GeoJSON`, `Shapefile` and `ArchGDAL`.

File: src/conversion.jl (hunk 2 above)
Comment (ErickChacon) [reply src/conversion.jl]: Yes, that would also works and make it cleaner.

File: src/conversion.jl [hunk 3]
```diff
@@ -40,33 +41,72 @@ GI.getfeature(::Any, d::Data, i) = d[i,:]
...
-  elseif gtrait isa GI.PolygonTrait
-    coords2poly(coords)
-  elseif gtrait isa GI.MultiPolygonTrait
-    coords2multipoly(coords)
+function GI.convert(::Type{Segment}, ::GI.LineTrait, geom)
+  p1, p2 = GI.startpoint(geom), GI.endpoint(geom)
+  if GI.is3d(geom)
+    Segment(Point(GI.x(p1), GI.y(p1), GI.z(p1)), Point(GI.x(p2), GI.y(p2), GI.z(p1)))
+  else
+    Segment(Point(GI.x(p1), GI.y(p1)), Point(GI.x(p2), GI.y(p2)))
```
Comment (juliohm) [inline src/conversion.jl:77]: can you comment why we can't use Segment(topoints(geom)) here like with the other geometries? The Segment should accept a vector of points as well.

Comment (ErickChacon) [reply src/conversion.jl:77]: Yes, it also works that way. Happy to modify that.

File: src/conversion.jl [hunk 4]
```diff
@@ -40,33 +41,65 @@ GI.getfeature(::Any, d::Data, i) = d[i,:]
...
 
-coords2multipoint(coords) = Multi(coords2point.(coords))
+GI.convert(::Type{Chain}, ::GI.LineStringTrait, geom) = tochain(geom, GI.is3d(geom))
 
-coords2chain(coords) = Chain(Point.(coords))
+GI.convert(::Type{Polygon}, ::GI.PolygonTrait, geom) = topolygon(geom, GI.is3d(geom))
 
-coords2multichain(coords) = Multi(coords2chain.(coords))
+function GI.convert(::Type{Multi}, ::GI.MultiPointTrait, geom)
+  Multi(topoints(geom, GI.is3d(geom)))
```
Comment (juliohm) [inline src/conversion.jl:79]: Should we follow the pattern in the other Multi methods and call is3d only once here as well?

**Review by juliohm — APPROVED** [review]

**Comment by ErickChacon** [comment]
All tests are passing in my machine. However, I am happy to add tests for the conversion using `ArchGDAL` and `GeometryBasics`. Any preference?

**Comment by juliohm** [comment]
Tests are failing for some HTTP reason, I think it is unrelated, can you confirm? We should probably add new tests as well with the stations.gpkg data set that originated the discussion in #13

**Comment by ErickChacon** [comment]
Great. Then, I will add tests for:

1. conversions from ArchGDAL to Meshes
2. read `stations.gpkg` using ArchGDAL loader 
3. read points, lines, and polygons? I already created these files to test this PR `.shp`, `.geojson` and `.gpkg`. Happy to add them to  `tests/data` folder.

**Comment by ErickChacon** [comment]
@juliohm As we are working with `Shapefile.jl`, `GeoJSON.jl` and `ArchGDAL.jl`. I added tests to verify that conversion from those packages to `Meshes.jl` works fine. Let me know if that is desirable for `GeoTables.jl` so I can continue to add test reading from files.

**Comment by juliohm** [comment]
That is great @ErickChacon , really helpful. Looking forward for the tests with files from disk.

**Comment by ErickChacon** [comment]
@juliohm, I completed this PR by adding tests for reading georeferenced files with `points`, `lines` and `polygons` using `Shapefile.jl`, `GeoJSON.jl`, and `ArchGDAL.jl`.

**Comment by ErickChacon** [comment]
I also wanted to test and check the possibility to write to other formats (e.g. using ArchGDAL), but that is probably better for a separated PR.

File: test/runtests.jl [hunk 5]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+
+      table = GeoTables.load(joinpath(datadir, "points.shp"))
+      @test Tables.schema(table).names == varnames
+      @test nitems(table) == 5
+      @test table.code[1] isa Integer
+      @test table.name[1] isa String
+      @test table.date[1] isa Date
+      @test table.variable[1] isa Real
+      @test table.geometry isa PointSet
+      @test table[1,:geometry] isa Point
```
Comment (juliohm) [inline test/runtests.jl:112]:
```suggestion
      @test table.geometry[1] isa Point
```

File: test/runtests.jl [hunk 6]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
-
-    table = GeoTables.load(joinpath(datadir,"ne_110m_land.shp"))
-    @test Tables.schema(table).names == (:featurecla, :scalerank, :min_zoom, :geometry)
-    @test nitems(table) == 127
-    @test all(==("Land"), table.featurecla)
-    @test all(∈([0,1]), table.scalerank)
-    @test all(∈([0.0,0.5,1.0,1.5]), table.min_zoom)
-    @test table.geometry isa GeometrySet
-    @test table[1,:geometry] isa Multi
+    @testset "custom datasets" begin
```
Comment (juliohm) [inline test/runtests.jl:72]:
```suggestion
    @testset "load" begin
```

File: test/runtests.jl [hunk 7]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
-    @test all(∈([0.0,0.5,1.0,1.5]), table.min_zoom)
-    @test table.geometry isa GeometrySet
-    @test table[1,:geometry] isa Multi
+    @testset "load" begin
+      table = GeoTables.load(joinpath(datadir,"path.shp"))
+      @test Tables.schema(table).names == (:ZONA, :geometry)
+      @test nitems(table) == 6
+      @test table.ZONA == ["PA 150","BR 364","BR 163","BR 230","BR 010","Estuarina PA"]
+      @test table.geometry isa GeometrySet
+      @test table[1,:geometry] isa Multi
```
Comment (juliohm) [inline test/runtests.jl:78]:
```suggestion
      @test table.geometry[1] isa Multi
```

File: test/runtests.jl [hunk 8]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+      table = GeoTables.load(joinpath(datadir,"zone.shp"))
+      @test Tables.schema(table).names == (:PERIMETER, :ACRES, :MACROZONA, :Hectares, :area_m2, :geometry)
+      @test nitems(table) == 4
+      @test table.PERIMETER == [5.850803650776888e6, 9.539471535859613e6, 1.01743436941e7, 7.096124186552936e6]
+      @test table.ACRES == [3.23144676827e7, 2.50593712407e8, 2.75528426573e8, 1.61293042687e8]
+      @test table.MACROZONA == ["Estuario","Fronteiras Antigas","Fronteiras Intermediarias","Fronteiras Novas"]
+      @test table.Hectares == [1.30772011078e7, 1.01411677447e8, 1.11502398263e8, 6.52729785685e7]
+      @test table.area_m2 == [1.30772011078e11, 1.01411677447e12, 1.11502398263e12, 6.52729785685e11]
+      @test table.geometry isa GeometrySet
+      @test table[1,:geometry] isa Multi
```
Comment (juliohm) [inline test/runtests.jl:89]:
```suggestion
      @test table.geometry[1] isa Multi
```

File: test/runtests.jl [hunk 9]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+
+      table = GeoTables.load(joinpath(datadir,"ne_110m_land.shp"))
+      @test Tables.schema(table).names == (:featurecla, :scalerank, :min_zoom, :geometry)
+      @test nitems(table) == 127
+      @test all(==("Land"), table.featurecla)
+      @test all(∈([0,1]), table.scalerank)
+      @test all(∈([0.0,0.5,1.0,1.5]), table.min_zoom)
+      @test table.geometry isa GeometrySet
+      @test table[1,:geometry] isa Multi
```
Comment (juliohm) [inline test/runtests.jl:98]:
```suggestion
      @test table.geometry[1] isa Multi
```

File: test/runtests.jl [hunk 10]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+      table = GeoTables.load(joinpath(datadir, "lines.shp"))
+      @test Tables.schema(table).names == varnames
+      @test nitems(table) == 5
+      @test table.code[1] isa Integer
+      @test table.name[1] isa String
+      @test table.date[1] isa Date
+      @test table.variable[1] isa Real
+      @test table.geometry isa GeometrySet
+      @test table[1, :geometry] isa Multi
+      @test collect(table[1, :geometry])[1] isa Chain
```
Comment (juliohm) [inline test/runtests.jl:123]:
```suggestion
      @test collect(table.geometry[1])[1] isa Chain
```

File: test/runtests.jl [hunk 11]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+
+      table = GeoTables.load(joinpath(datadir, "lines.shp"))
+      @test Tables.schema(table).names == varnames
+      @test nitems(table) == 5
+      @test table.code[1] isa Integer
+      @test table.name[1] isa String
+      @test table.date[1] isa Date
+      @test table.variable[1] isa Real
+      @test table.geometry isa GeometrySet
+      @test table[1, :geometry] isa Multi
```
Comment (juliohm) [inline test/runtests.jl:122]:
```suggestion
      @test table.geometry[1] isa Multi
```

File: test/runtests.jl [hunk 12]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+      table = GeoTables.load(joinpath(datadir, "polygons.shp"))
+      @test Tables.schema(table).names == varnames
+      @test nitems(table) == 5
+      @test table.code[1] isa Integer
+      @test table.name[1] isa String
+      @test table.date[1] isa Date
+      @test table.variable[1] isa Real
+      @test table.geometry isa GeometrySet
+      @test table[1, :geometry] isa Multi
+      @test collect(table[1, :geometry])[1] isa PolyArea
```
Comment (juliohm) [inline test/runtests.jl:134]:
```suggestion
      @test table.geometry[1] isa Multi
      @test collect(table.geometry[1])[1] isa PolyArea
```

File: test/runtests.jl [hunk 13]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+
+      table = GeoTables.load(joinpath(datadir, "points.geojson"))
+      @test isempty(setdiff(Tables.schema(table).names, varnames))
+      @test nitems(table) == 5
+      @test table.code[1] isa Integer
+      @test table.name[1] isa String
+      @test table.date[1] isa String
+      @test table.variable[1] isa Real
+      @test table.geometry isa PointSet
+      @test table[1,:geometry] isa Point
```
Comment (juliohm) [inline test/runtests.jl:148]:
```suggestion
      @test table.geometry[1] isa Point
```

File: test/runtests.jl [hunk 14]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+
+      table = GeoTables.load(joinpath(datadir, "lines.geojson"))
+      @test isempty(setdiff(Tables.schema(table).names, varnames))
+      @test nitems(table) == 5
+      @test table.code[1] isa Integer
+      @test table.name[1] isa String
+      @test table.date[1] isa String
+      @test table.variable[1] isa Real
+      @test table.geometry isa GeometrySet
+      @test table[1, :geometry] isa Chain
```
Comment (juliohm) [inline test/runtests.jl:158]:
```suggestion
      @test table.geometry[1] isa Chain
```

File: test/runtests.jl [hunk 15]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+
+      table = GeoTables.load(joinpath(datadir, "polygons.geojson"))
+      @test isempty(setdiff(Tables.schema(table).names, varnames))
+      @test nitems(table) == 5
+      @test table.code[1] isa Integer
+      @test table.name[1] isa String
+      @test table.date[1] isa String
+      @test table.variable[1] isa Real
+      @test table.geometry isa GeometrySet
+      @test table[1, :geometry] isa PolyArea
```
Comment (juliohm) [inline test/runtests.jl:168]:
```suggestion
      @test table.geometry[1] isa PolyArea
```

File: test/runtests.jl [hunk 16]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+
+      table = GeoTables.load(joinpath(datadir, "points.gpkg"))
+      @test Tables.schema(table).names == varnames
+      @test nitems(table) == 5
+      @test table.code[1] isa Integer
+      @test table.name[1] isa String
+      @test table.date[1] isa DateTime
+      @test table.variable[1] isa Real
+      @test table.geometry isa PointSet
+      @test table[1,:geometry] isa Point
```
Comment (juliohm) [inline test/runtests.jl:182]:
```suggestion
      @test table.geometry[1] isa Point
```

File: test/runtests.jl [hunk 17]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+
+      table = GeoTables.load(joinpath(datadir, "lines.gpkg"))
+      @test Tables.schema(table).names == varnames
+      @test nitems(table) == 5
+      @test table.code[1] isa Integer
+      @test table.name[1] isa String
+      @test table.date[1] isa DateTime
+      @test table.variable[1] isa Real
+      @test table.geometry isa GeometrySet
+      @test table[1, :geometry] isa Chain
```
Comment (juliohm) [inline test/runtests.jl:192]:
```suggestion
      @test table.geometry[1] isa Chain
```

File: test/runtests.jl [hunk 18]
```diff
@@ -2,40 +2,205 @@ using GeoTables
...
+
+      table = GeoTables.load(joinpath(datadir, "polygons.gpkg"))
+      @test Tables.schema(table).names == varnames
+      @test nitems(table) == 5
+      @test table.code[1] isa Integer
+      @test table.name[1] isa String
+      @test table.date[1] isa DateTime
+      @test table.variable[1] isa Real
+      @test table.geometry isa GeometrySet
+      @test table[1, :geometry] isa PolyArea
```
Comment (juliohm) [inline test/runtests.jl:202]:
```suggestion
      @test table.geometry[1] isa PolyArea
```

**Review by juliohm — APPROVED** [review]

**Comment by juliohm** [comment]
Amazing work @ErickChacon, I will run the tests locally and merge it. I think we are ready for GeoTables.jl v1.0 👍🏽

**Pull request merged**